src/gas-price/gas-price.ts

```diff
@@ -0,0 +1,205 @@
+  recommendedGasPriceMultiplier: number,
+  maxScalingMultiplier: number,
+  lag: number,
+  scalingWindow: number
+) =>
+  Math.min(
+    recommendedGasPriceMultiplier + (maxScalingMultiplier - recommendedGasPriceMultiplier) * (lag / scalingWindow),
+    maxScalingMultiplier
+  );
+
+/**
+ * Fetches the provider recommended gas price and saves it in the store.
+ * @param chainId
+ * @param providerName
+ * @param rpcUrl
+ * @returns {ethers.BigNumber}
+ */
+export const updateGasPriceStore = async (chainId: string, providerName: string, rpcUrl: string) => {
+  const provider = new ethers.providers.StaticJsonRpcProvider(rpcUrl, {
+    chainId: Number.parseInt(chainId, 10),
+    name: chainId,
+  });
+  // Get the provider recommended gas price
+  const gasPrice = await provider.getGasPrice();
+  // Save the new provider recommended gas price to the state
+  setStoreGasPrices(chainId, providerName, gasPrice);
+
+  return gasPrice;
+};
+
+/**
+ * Fetches the provider recommended gas price, saves it in the store and clears out expired gas prices.
+ * @param chainId
+ * @param providerName
+ * @param rpcUrl
+ */
+export const gasPriceCollector = async (
+  chainId: string,
+  providerName: string,
+  rpcUrl: string,
+  sanitizationSamplingWindow: number
+) => {
+  // Initialize the gas store for the chain if not already present
+  initializeGasStore(chainId, providerName);
+  clearExpiredStoreGasPrices(chainId, providerName, sanitizationSamplingWindow);
+  await updateGasPriceStore(chainId, providerName, rpcUrl);
+};
+
+/**
+ *  Calculates the gas price to be used in a transaction based on sanitization and scaling settings.
+ * @param chainId
+ * @param providerName
+ * @param rpcUrl
+ * @param gasSettings
+ * @param nonce
+ * @returns {ethers.BigNumber}
+ */
+export const getAirseekerRecommendedGasPrice = async (
+  chainId: string,
+  providerName: string,
+  rpcUrl: string,
+  gasSettings: GasSettings,
+  newDataFeedUpdateOnChainValues?: {
+    dataFeedId: string;
+    newDataFeedValue: DataFeedValue;
+  }
+): Promise<ethers.BigNumber> => {
+  const { recommendedGasPriceMultiplier, sanitizationPercentile, scalingWindow, maxScalingMultiplier } = gasSettings;
+  // Get the configured percentile of historical gas prices before adding the new price
+  const percentileGasPrice = getPercentile(
+    sanitizationPercentile,
+    gasPriceStore[chainId]![providerName]!.gasPrices.map((gasPrice) => gasPrice.price)
+  );
+
+  const gasPrice = await updateGasPriceStore(chainId, providerName, rpcUrl);
+
+  const lastDataFeedValue =
+    newDataFeedUpdateOnChainValues &&
+    gasPriceStore[chainId]![providerName]!.lastOnChainDataFeedValues[newDataFeedUpdateOnChainValues.dataFeedId];
+  // Check if the next update is a retry of a pending transaction and if it has been pending longer than scalingWindow
+  if (
+    lastDataFeedValue &&
+    newDataFeedUpdateOnChainValues &&
+    lastDataFeedValue?.value === newDataFeedUpdateOnChainValues.newDataFeedValue.value &&
```

Comment:
In this [slack comment](https://api3workspace.slack.com/archives/C05TQPT7PNJ/p1698222551839029?thread_ts=1698220736.112239&cid=C05TQPT7PNJ) I keep wondering about this. 

We will be making new transactions all the time (since the singed data changes asynchronously) and this will not track pending txs correctly. I do not see any other way except relying on wallet nonces or awaiting the transaction promise.

Reply:
I guess the important part is that if the on-chain value changes there had to be an update. If that's only what we care about then the logic is sound.